### PR TITLE
PR for code review

### PR DESCRIPTION
**Pull request by joegoldbeck**

### CODE REVIEW

File: src/utils/create.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    }
+
+    if (parent) {
+        parent.appendChild(element)
+    }
+
+    if (dataAttr.length) {
+        dataAttr.forEach(([ attrName, attrValue]) =>  {
+            if (attrValue === '') element.setAttribute(attrName, '')
+            if (attrName.match(/value|type|id|placeholder|cols|rows|autocorrect|spellcheck|role|aria-checked|for|title/)) {
```
Comment (joegoldbeck) [inline src/utils/create.js:24]: seems if you're not going to early return this should just be `if` `elif` `else`

alternatively, early return all of them

```
if (attrValue === '') {
    return element.setAttribute(attrName, '')
}
if (attrName.match(...)){
   return ...
}

return element.dataset[attrName] = attrValue
```

File: src/utils/words-and-translations.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
+export default function allWords() {
+    const allWords = {};
```
Comment (joegoldbeck) [inline src/utils/words-and-translations.js:2]: Personal style: in general, if I'm building an object and then returning it, I like to just do it all at once unless it's too complicated to read, since it make the intention immediately obvious. Also fewer places for bugs to hide. Any errors that happen more likely syntax errors than runtime errors.

```
export default function allWords() {
  return {
     en: {
         actionA: ['cry', 'dance', 'dive', 'draw', 'fish', 'fly', 'hug', 'jump'],
         actionB: ['open', 'play', 'point', 'ride', 'run', 'sing', 'skip', 'swim'],
         actionC: ['argue', 'puild', 'carry', 'catch', 'drive', 'drop', 'pull', 'push'],
         adjectives: ['big', 'small', 'fast', 'slow', 'friendly', 'unfriendly', 'young', 'old'],
         animalB: ['bird', 'fish', 'frog', 'giraffe', 'lion', 'mouse', 'turtle', 'dolphin'],
         animalA: ['cat', 'chick', 'chicken', 'dog', 'horse', 'pig', 'rabbit', 'sheep'],
         emotions: ['sad', 'angry', 'happy', 'tired', 'surprised', 'scarred', 'smile', 'laugh'],
         clothes: ['skirt', 'pants', 'blouse', 'dress', 'boots', 'shirt', 'coat', 'shoes']
     },
     ru: {
         actionA: ['плакать', 'танцевать', 'нырять', 'рисовать', 'рыбачить', 'летать', 'обнимать', 'прыгать'],
         actionB: ['открывать', 'играть', 'указывать', 'ездить', 'бегать', 'петь', 'скакать', 'плавать'],
         actionC: ['спрорить', 'строить', 'переносить', 'ловить', 'водить', 'бросать', 'тянуть', 'толкать'],
         adjectives: ['большой', 'маленький', 'быстрый', 'медленный', 'дружелюбный', 'недружелюбный', 'молодой', 'старый'],
         animalB: ['птица', 'рыба', 'лягушка', 'жираф', 'лев', 'мышь', 'черепаха', 'дельфин'],
         animalA: ['кот', 'курица', 'цыпленок', 'собака', 'лошадь', 'свинья', 'кролик', 'овца'],
         emotions: ['грустный', 'сердитый', 'счастливый', 'усталый', 'удивленный', 'испуганный', 'улыбающийся', 'смеющийся'],
         clothes: ['юбка', 'брюки', 'блузка', 'платье', 'ботинки', 'рубашка', 'пальто', 'туфли']
     }
   }
 }

File: src/utils/create.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
+export default function create(el, classNames, child, parent, ...dataAttr) {
```
Comment (joegoldbeck) [inline src/utils/create.js:1]: Seems odd to be building your own function for this. Certainly fine for a learning exercise! But for anything serious I'd expect to have well-tested open-source tooling or a framework to handle this :)

File: src/utils/category-images.js [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    cat4.src = animalA;
+    allImages.catImg.push(cat4);
+    allImages.catName.push('animals A');
+
+    const cat5 = new Image;
+    cat5.src = animalB;
+    allImages.catImg.push(cat5);
+    allImages.catName.push('animal B');
+
+    const cat6 = new Image;
```
Comment (joegoldbeck) [inline src/utils/category-images.js:125]: Lots of repetition here. Looks like a good opportunity for a helper function, or new abstraction, and potentially a `map` as well

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map

File: src/utils/category-images.js [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        actionA : [cat11,cat12,cat13,cat14,cat15,cat16,cat17,cat18],
+        actionB : [cat21,cat22,cat23,cat24,cat25,cat26,cat27,cat28],
+        actionC : [cat31,cat32,cat33,cat34,cat35,cat36,cat37,cat38],
+        adjectives : [cat41,cat42,cat43,cat44,cat45,cat46,cat47,cat48],
+        animalA : [cat51,cat52,cat53,cat54,cat55,cat56,cat57,cat58],
+        animalB : [cat61,cat62,cat63,cat64,cat65,cat66,cat67,cat68],
+        clothes : [cat71,cat72,cat73,cat74,cat75,cat76,cat77,cat78],
+        emotions : [cat81,cat82,cat83,cat84,cat85,cat86,cat87,cat88]
+    };
+
```
Comment (joegoldbeck) [inline src/utils/category-images.js:98]: Pretty opaque what is happening here!

File: src/utils/audio.js [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+import skip from '../assets/sounds/skip.mp3';
+import skirt from '../assets/sounds/skirt.mp3';
+import smile from '../assets/sounds/smile.mp3';
+import success from '../assets/sounds/success.mp3';
+import surprised from '../assets/sounds/surprised.mp3';
+import swim from '../assets/sounds/swim.mp3';
+import tired from '../assets/sounds/tired.mp3';
+import turtle from '../assets/sounds/turtle.mp3';
+
+export default function getSound() {
```
Comment (joegoldbeck) [inline src/utils/audio.js:53]: Nit: I feel like there should be a way to make this more compact.

One possibility to apply the `new Audio` at the end rather than repeating so many times could be https://lodash.com/docs/#map

Really not a big deal and maybe not worth the indirection since each line is so simple but just wanted to point it out

File: src/start.js [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+            currWordCards.push(newWordCard(img, en, ru, en));
+        }
+        //flip the word card
+        _.shuffle(currWordCards).forEach( card => {
+            main.appendChild(card.domElement);
+            card.domElement.addEventListener('click', e => {
+                if(e.target.classList.contains('card-flip')) {
+                    card.domElement.classList.add('flipped');
+                }
+                if(e.target.classList.contains('wordCard') && !e.target.classList.contains('flipped')) {
```
Comment (joegoldbeck) [inline src/start.js:59]: Man, you're really going vanilla javascript here XD. 

Again, for a small learning project, why not! But for anything serious, a framework would tighten this up so don't need some many dots `___.___.___`.

File: src/start.js [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+import addTopAppBar from './components/mdc-top-app-bar/index.js';
+import addButton from './components/mdc-button/index.js';
+import newCategory from './components/category/index.js';
+import images from './utils/category-images.js';
+import newWordCard from './components/word-card/index.js';
+import allWords from './utils/words-and-translations.js';
+import getSound from './utils/audio.js';
+import _ from 'lodash';
+
+export default function start() {
```
Comment (joegoldbeck) [inline src/start.js:12]: A little hard to follow! At this level of app complexity it's probably fine, but at a higher level of complexity it would bite you for sure. This is another place where a framework may provide some additional structure for the "wire all the things together" part.

Avoiding spaghetti code (which this isn't currently, but risks becoming!) is pretty crucial

File: package.json [hunk 9]
```diff
@@ -0,0 +1,47 @@
+{
+  "name": "english-for-kids",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
+  "private": "true",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1",
```
Comment (joegoldbeck) [inline package.json:8]: may consider to add some unit tests at some point. In general, I find that writing good tests forces me to write maintain reasonable abstractions and that the testable chunks I have to break it up into in order to write reasonable tests results in maintainable/understandable chunks as well

**Review by joegoldbeck — COMMENTED** [review]
This is neat! Really impressed what you've done without any framework and pretty much just pure js. Left a handful of initial comments and more than happy to do further rounds or discuss anything!

**Review by joegoldbeck — COMMENTED** [review]
Whoops I just saw your actual question

> On this task I practiced how to divide applications into components. So, my main question is what I did wrong and how to do it better.

I think the comments on framework and testing still both hold. In general, I'd want both of those in a real application and they are both helpful with breaking into components. 

Break-up seems mostly ok at first pass, though `start` function is rather noisy. If possible, I'd pull more stuff (for ex `gameBoard`) out of there so that that function is more just calling setup functions on the necessary components.

I'll also sleep on it and see if I have any additional thoughts specifically about how to break things up :)

File: package.json (hunk 9 above)
Comment (llama-with-thumbs) [reply package.json:8]: Can you recommend any particular library for tests? Mocha or Chai, maybe both.

File: src/start.js (hunk 7 above)
Comment (llama-with-thumbs) [reply src/start.js:59]: 100% agree, __.__.__ - is really hart to maintain. it's like juggle - when you dropped something you need to start over.

File: src/components/category/index.js [hunk 10]
```diff
@@ -0,0 +1,14 @@
+import create from '../../utils/create.js';
+
+export default function newCategory(categoryName, image, dataName) {
+    const category = {};
+
+    const categoryNameElement = create('div', 'category__name', categoryName, null);
+    const cartegoryElement = create( 'div', 'category mdc-ripple-surface' , categoryNameElement, null, ['name', dataName]);
```
[the remark this answers is not in the document]
Comment (llama-with-thumbs) [reply src/components/category/index.js:7]: Thank you. I installed Code Spell Checker for my Visual Studio Code.
I had Prettier, but it was poorly set up - so I took some time and did it a little better.

File: package.json (hunk 9 above)
Comment (joegoldbeck) [reply package.json:8]: I've used mocha a fair bit, sometimes chai for assertions. Mostly for backend myself, but I do think can be used for some frontend stuff

Jest is pretty popular for frontend, I think more fully featured in that direction https://jestjs.io

File: src/components/mdc-drawer/index.js [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+import {MDCDrawer} from "@material/drawer";
+
+
+export default function addDrawer() {
+  const listEl = document.querySelector('.mdc-drawer .mdc-list');
+  const drawer = MDCDrawer.attachTo(document.querySelector('.mdc-drawer'));
+  const closeDrawerButton = document.getElementById("close-drawer");
+
+
+    setTimeout(function() {
```
[the remark this answers is not in the document]
Comment (llama-with-thumbs) [reply src/components/mdc-drawer/index.js:12]: I made it as temporary solution... but obviously forgot to make refactoring.

[the remark this answers is not in the document]
Comment (llama-with-thumbs) [reply src/components/mdc-drawer/index.js:12]: What is best practice when you want to do some actions with not yet added DOM element? ...Collbacks?

File: src/utils/words-and-translations.js (hunk 2 above)
Comment (llama-with-thumbs) [reply src/utils/words-and-translations.js:2]: It was my shllow

File: src/utils/create.js (hunk 3 above)
Comment (llama-with-thumbs) [reply src/utils/create.js:1]: You right, I made it for learning purposes only.

File: src/utils/create.js [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+    let element = null
+    try {
+        element = document.createElement(el)
+    } catch (error) {
+        throw new Error('unable Tag name')
```
[the remark this answers is not in the document]
Comment (llama-with-thumbs) [reply src/utils/create.js:6]: You right, thank you. 
What error message should contain besides well formulated description of particular problem? I mean, is there any pro tips on that?:)))

File: src/utils/category-images.js (hunk 5 above)
Comment (llama-with-thumbs) [reply src/utils/category-images.js:98]: I'd better named images something like 'img11', 'img12'.
The first number is a category index and the second is position incide category. 
Didn't find batter pattern.

File: src/utils/audio.js (hunk 6 above)
Comment (llama-with-thumbs) [reply src/utils/audio.js:53]: I started to rewrite the sound part, but it is a mess. So I did what I planed for the very beginning - use library for voice generation.
`import speakTheWord from './utils/speakTheWord.js';
``speakTheWord(e.target.dataset.english);`

File: src/components/mdc-drawer/index.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (joegoldbeck) [reply src/components/mdc-drawer/index.js]: I don't know the right way to express this in this particular case, partly because I'm not sure _what_ it's waiting for. (This is one downside to the setTimeout trick, it doesn't really codify why there is a wait, and so can easily become broken) 

Callbacks (or promises) are probably the right key here, ideally you'd want one that will resolve / be called when the desired component has been added to DOM. Typically in a framework this would be handled. Perhaps there would be some "onRender" or "onReady" callback, or even more likely, you might register the components with the framework in some hierarchy, and the framework would take care of insertion order. 

It is rare in modern frontend development to have to have to deal with this sequencing directly, rather the framework will provide the tools so you don't need to build the plumbing from ground up. Very non-trivial not to get spaghetti code if building from scratch.

File: src/utils/create.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (joegoldbeck) [reply src/utils/create.js]: Two changes

1) More clear error message itself ("unable Tag name" is pretty unclear. seems like "failed to create element" is may be a better message. 
2) Log the info needed to debug. Here, we don't know _what_ element failed to render, _and_ the error itself is swallowed so entirely unclear why it failed.

I might be tempted to do something like

```
console.error("Unable to create element", el)
throw error
```

File: src/utils/category-images.js [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+    'actionB',
+    'actionC',
+    'animalA',
+    'animalB',
+    'adjectives',
+    'emotions',
+    'clothes',
+  ];
+
+  for (let i = 0; i < allImages.categoryNameList.length; i++) {
```
Comment (joegoldbeck) [inline src/utils/category-images.js:120]: Something feels odd to me about how all this is getting built, and as a result you end up with lots of `for` loops, rather than `map`s for ex. 

In particular, it looks like you need to ensure that `allImages.categoryThumbnailList` is the same length and alignment as `allImages.categoryNameList` etc... It feels odd to be maintaining two arrays that must be the same length, rather than a single array of objects! I haven't spent enough time to determine exactly how to link them, but I encourage you to construct these as if you are building one list of instances, rather than a few lists that you need to slice across in order to construct details of a single thing.

This becomes clear up top in your definitions. You are defining the same order multiple times. This is represents a maintenance burden / risk for example: if you want to add a new item to the list, you need to add it in the same spot in each list otherwise you will have a weird bug.

Each object in this list might be like

```
{
  "name": "actionA",
  "img": actionA,
  "innerImg": [img11, img12, img13, img14, img15, img16, img17, img18],
}
```

File: src/utils/category-images.js [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+    actionB: [img21, img22, img23, img24, img25, img26, img27, img28],
+    actionC: [img31, img32, img33, img34, img35, img36, img37, img38],
+    adjectives: [img41, img42, img43, img44, img45, img46, img47, img48],
+    animalA: [img51, img52, img53, img54, img55, img56, img57, img58],
+    animalB: [img61, img62, img63, img64, img65, img66, img67, img68],
+    clothes: [img71, img72, img73, img74, img75, img76, img77, img78],
+    emotions: [img81, img82, img83, img84, img85, img86, img87, img88],
+  };
+
+  allImages.categoryNameList = [
```
Comment (joegoldbeck) [inline src/utils/category-images.js:109]: Style/naming: in general, I prefer to use plural as opposed to saying "list", for simultaneous compactness readability. So from a naming perspective this could be `allImages.categoryNames`.

(Though see below I think it's a big odd to separate all these out versus just one list of objects representing `categories`

File: src/utils/speakTheWord.js [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+        voice: 'Google UK English Male',
+      })
+      .then((data) => {
+        speech.speak({
+          text: text,
+          queue: false,
+        });
+      })
+      .catch((e) => {
+        console.error('An error occurred while initializing : ', e);
```
Comment (joegoldbeck) [inline src/utils/speakTheWord.js:20]: Can be more explicit about what's up here:

`console.error('An error occurred while initializing speech : ', e);`

File: src/utils/speakTheWord.js [hunk 16]
```diff
@@ -0,0 +1,26 @@
+import Speech from 'speak-tts';
+function speakTts(text) {
+  function _init() {
```
Comment (joegoldbeck) [inline src/utils/speakTheWord.js:3]: In this case, it's unclear to me why you define this function and then immediately call it.

File: src/utils/speakTheWord.js [hunk 17]
```diff
@@ -0,0 +1,26 @@
+import Speech from 'speak-tts';
+function speakTts(text) {
+  function _init() {
+    const speech = new Speech();
+    speech
```
Comment (joegoldbeck) [inline src/utils/speakTheWord.js:5]: If you have a promise chain, you should generally `return` it. e.g. `return speech.init(...) .....` Then if the caller wants to continue the promise chain in order to do something _after_ the speech happens, it can. 

Of course, you'll also need to either unwrap this from the `_init` function, or `return _init()` as well. (But again, I don't see what `_init()` is giving you here, I think you can just remove that fn at return this directly`

File: src/utils/speakTheWord.js [hunk 18]
```diff
@@ -0,0 +1,26 @@
+import Speech from 'speak-tts';
```
Comment (joegoldbeck) [inline src/utils/speakTheWord.js:1]: neat!

File: src/utils/audio.js (hunk 6 above)
Comment (joegoldbeck) [reply src/utils/audio.js]: Good call! Sounds much easier to maintain
